utils/s2n_blob.c

```diff
@@ -47,3 +47,47 @@ int s2n_blob_char_to_lower(struct s2n_blob *b)
 
     return 0;
 }
+
+/* inverse an ascii value to a hexidecimal value
+ * accounts for all possible char values, where 255 is invalid value */
+static const uint8_t hex_inverse[256] = {
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+      0,   1,   2,   3,   4,   5,   6,   7,   8,   9, 255, 255, 255, 255, 255, 255,
+    255,  10,  11,  12,  13,  14,  15, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255,  10,  11,  12,  13,  14,  15, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255
+};
+
+/* takes a hex string and writes values in the s2n_blob
+ * string needs to a valid hex and blob needs to be large enough */
+int s2n_hex_string_to_bytes(const char *str, struct s2n_blob *blob)
+{
+    notnull_check(str);
+    notnull_check(blob);
+    /* protects against overflows */
+    gte_check(blob->size, strlen(str) / 2);
+    S2N_ERROR_IF(strlen(str) % 2 != 0, S2N_ERR_INVALID_HEX);
+
+    for (int i = 0; i < strlen(str); i += 2) {
```

Comment:
This code ends up calling strlen  n/2 + 2 times but the value never changes. Might be better as a variable!

Reply:
I’m sure the compiler would detect it as invariant in the loop and optimize it away, but making a variable makes this far more obvious to the casual reader